pkg/checker/functools.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2017-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package checker
+
+import (
+	"fmt"
+	"reflect"
+)
+
+func makeCheckRunner(f interface{}, args ...interface{}) (CheckRunner, error) {
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `CheckRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+func makeRemediationRunner(f interface{}, args ...interface{}) (RemediationRunner, error) {
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `RemediationRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
```

Comment:
This function is the same as `makeCheckRunner`. Do we need to duplicate this code?

Reply:
yes and no - let me explain. These two functions return respectively `CheckRunner` and `RemediationRunner`. They just happen to have the same signature, but this is not guaranteed, and they are logically distinct. So I kept them separate and only refactored out  the code that now lives in `validateFunction`.

I'd suggest two options:
* I keep the duplication. Pros: the functions remain logically independent. Cons: code duplication is not nice (but it's a small amount)
* I create another function `makeFunctionThatReturnsError` (whoops..) with the same body of these two functions, but I wrap them under a different name and signature. In other words:

```
func makeFunctionThatReturnsError(f interface{}, args ...interface{}) func() error {
    ...
}

func makeCheckRunner(f interface{}, args ...interface{}) RemediationRunner {
    return makeFunctionThatReturnsError(f, args...)
}

func makeRemediationRunner(f interface{}, args ...interface{}) RemediationRunner {
    return makeFunctionThatReturnsError(f, args...)
}
```

What do you suggest?

Reply:
meanwhile I'll go with the latter

Reply:
Personally I would do the second option. But, I'm a little-bit crazy about deduplication, do not take it seriously :)

Reply:
I changed it already, with a longer function name